File: Algorithm/WordBreak.java
// // Solution: DP
// // Time complexity: O(n^2)
// // Space complexity: O(n^2)
// class Solution {
//     public boolean wordBreak(String s, List<String> wordDict) {
//         // Create a hashset of words for fast query.
//         Set<String> dict = new HashSet<String>(wordDict);
//         // Query the answer of the original string.
//         Map<String, Boolean> mem = new HashMap<String, Boolean>();
//         return workBreak(s, mem, dict);
//     }
    
//     private boolean workBreak(String s, Map<String, Boolean> mem, Set<String> dict) {
//         // In memory. directly return.
//         if (mem.containsKey(s)) return mem.get(s);
//         // Whose string is a word, memorize and return.
//         if (dict.contains(s)) {
//             mem.put(s, true);
//             return true;
//         }
//         // Try every break point.
//         for (int i = 1; i < s.length(); ++i) {
//             // Find the solution for s.
//             if (dict.contains(s.substring(i)) && workBreak(s.substring(0, i), mem, dict)) {
//                 mem.put(s, true);
//                 return true;
//             }
//         }
//         // No solution for s, memorize and return.
//         mem.put(s, false);
//         return false;
//     }
// }

// Solution: without using dict. 
class Solution {
    private Map<String, Boolean> mem_;
    public boolean wordBreak(String s, List<String> wordDict) {
        // Mark evert word as breakable.
        mem_ = new HashMap<>();
        for (String word : wordDict)
            mem_.put(word, true);

        // Query the answer of the original string.
        return wordBreak(s);
    }
    
    private boolean wordBreak(String s) {
        // In memory, directly return.
        if (mem_.containsKey(s)) return mem_.get(s);
        
        // Try every break point.
        for (int j = 1; j < s.length(); ++j) {
            Boolean it = mem_.get(s.substring(j));  // 这里注意接受it用Boolean类型，而不是boolean。
            // Find the solution for s.
            if (it != null && it && wordBreak(s.substring(0, j))) {
                mem_.put(s, true);
                return true;
            }
        }
        
        // No solution for s, memorize and return.
        mem_.put(s, false);
        return false;
    }
}

File: Algorithm/WordBreak.cpp
// // Solution: DP
// // Time complexity: O(n^2)
// // Space complexity: O(n^2)
// class Solution {
// public:
//     bool wordBreak(string s, vector<string>& wordDict) {
//         // Create a hashset of words for fast query.
//         unordered_set<string> dict(wordDict.cbegin(), wordDict.cend());
//         // Query the answer of the original string.
//         return wordBreak(s, dict);
//     }
    
//     bool wordBreak(const string& s, const unordered_set<string>& dict) {
//         // In memory, directly return.
//         if (mem_.count(s)) return mem_[s];
//         // Whole string is a word, memorize and return.
//         if (dict.count(s)) return mem_[s] = true;
//         // Try every break point.
//         for (int j = 1; j < s.length(); ++j) {
//             const string left = s.substr(0, j);
//             const string right = s.substr(j);
//             // Find the solution for s.
//             if (dict.count(right) && wordBreak(left, dict))
//                 return mem_[s] = true;
//         }
//         // No solution for s, memorize and return.
//         return mem_[s] = false;
//     }
// private:
//     unordered_map<string, bool> mem_;
// };

// C++ V2 without using dict.
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        // Mark every word as breakable.
        for (const string& word : wordDict)
            mem_.emplace(word, true);
        
        // Query the answer of the original string.
        return wordBreak(s);
    }
    
    bool wordBreak(const string& s) {
        // In memory, directly return.
        if (mem_.count(s)) return mem_[s];
        
        // Try every break point.
        for (int j = 1; j < s.length(); ++j) {
            auto it = mem_.find(s.substr(j));
            // Find the solution for s.
            if (it != mem_.end() && it->second && wordBreak(s.substr(0, j)))
                return mem_[s] = true;
        }
        
        // No solution for s, memorize and return.
        return mem_[s] = false;
    }
private:
    unordered_map<string, bool> mem_;
};
